Add tests for warnings when passing block to creation of new IO object

herwinw opened the pull request:
I think these were all `IO` subclasses, which means I probably missed half of them.
I used the syntax `IO::open` instead of `IO.open` in the descriptions of the tests, since that is what the warnings say.

herwinw commented:
There is a change in the TCP server spec where I have no idea how it has gotten into this pull request, I will fix that tomorrow

andrykonchin commented:
Looks good.

It seems this case is missing also for `TCPSocket` and `StringIO` classes

herwinw commented:
https://github.com/ruby/spec/blob/master/library/stringio/new_spec.rb

`StringIO` already has this test (which I did not know until 10 seconds agon).

herwinw commented:
Added `TCPSocket` (the specs for `new` are in `initialize_spec`, that's why I missed it at first glance). I updated the `StringIO` spec to match the style of the other ones, and added a `raise` in the block to verify that the block is not used.

eregon commented on library/socket/unixserver/new_spec.rb:
```diff
@@ -2,5 +2,13 @@
 
 describe "UNIXServer.new" do
-  it_behaves_like :unixserver_new, :new
+  platform_is_not :windows do
```
Could you move the `platform_is_not` outside? Because UNIXServer is not even defined on Windows (actually it is in some cases but we are ignoring that so far to keep things simple). Same below.

herwinw replied:
I extended this one a little bit to keep things consistent:
* Replace `platform_is_not :windows` with `with_feature :unix_socket`, which looks a bit more generic. The current code had both in use. This could resolve the "in some cases on Windows", dependign on the implementation of `with_feature`.
* Move these guards to the outer block in all spec files, and remove them from the shared files.

eregon replied:
Great, that's even better, thank you!
